udata/api/cache.py

```diff
@@ -0,0 +1,124 @@
+    on_discussion_deleted
+)
+
+
+ONE_MINUTE = 60
+ONE_HOUR = ONE_MINUTE * 60
+ONE_DAY = ONE_HOUR * 24
+ONE_WEEK = ONE_DAY * 7
+ONE_MONTH = ONE_DAY * 30
+ONE_YEAR = ONE_DAY * 365
+
+
+def cache_page(class_instance, check_serverside=True, personal=False,
+               client_timeout=0, server_timeout=ONE_HOUR, key='view-cache%s',
+               x_cache_header=True, serializer=lambda a: a):
+    """An attempt to make the ultimate Flask cache decorator.
+
+    Differs from the official Flask Caching Decorator
+    http://flask.pocoo.org/docs/0.12/patterns/viewdecorators/#caching-decorator
+    by dealing with appropriated headers for server AND client-side caches.
+
+    Key points:
+    * Generates the cache key using the full path (using the query string).
+    * Uses `Cache-Control`, `ETag` and `Last-Modified` headers.
+    * Respects browser's hard-refresh.
+
+    Parameters:
+    * `check_serverside`: either the client verifies with the server if
+      its cached versions are still valid or not.
+    * `personal`: set to `True` if the response contains sensitive
+      informations only pertinent for a given user.
+    * `client_timeout`: time in seconds before the client considers the
+      cache as being invalid.
+    * `server_timeout`: time in seconds before the server considers the
+      cache as being invalid.
+    * `key`: you can customize the cache key, it must contains `%s`
+      which will be replaced by the full path (including query string).
+    * `x_cache_header`: either put a `X-Cache` header to know which
+      cache is being hit or not. Useful to debug but not required.
+    * `serializer`: if the response from your view needs to be
+      serialized prior to be converted to a Flask Response object.
+
+    Inspirations:
+    * http://codereview.stackexchange.com/questions/147038/↩︎
+      improving-the-flask-cache-decorator
+    * https://gist.github.com/glenrobertson/954da3acec84606885f5
+
+    Resources:
+    * https://jakearchibald.com/2016/caching-best-practices/
+    * https://developers.google.com/web/fundamentals/performance/↩︎
+      optimizing-content-efficiency/http-caching
+
+    Image:
+    * https://developers.google.com/web/fundamentals/performance/↩︎
+      optimizing-content-efficiency/images/http-cache-decision-tree.png
+    """
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            # Before you think the logic is reversed, read Jake's article.
+            cache_policy = check_serverside and 'no-cache' or 'must-revalidate'
+            if personal:
+                cache_policy += ', private'
+            else:
+                cache_policy += ', public'
+            cache_policy += ', max-age={}'.format(client_timeout)
+            headers = {}
+            headers['Cache-Control'] = cache_policy
+            client_etag = request.headers.get('If-None-Match')
+            cache_key = key % request.full_path  # Include querystring.
```

Comment:
Sadly this will generate an infinite of cache entries never reused on the admin part because to ensure there is not cache, each request is sent with`_=<timestamp>` query string.
I think we need a finer cache key generation mecanism. Maybe something like `f(endpoint, method, request.view_args, request.values)`
In falsk-restplus, it would be easier because we have the `@api.expect` decorator which is listing the varying input content so cache kay would be `f(endpoint, method, request.view_args, parsed exectations)`